connectors/sources/gitlab/client.py

```diff
@@ -0,0 +1,1029 @@
+            retry_after = response.headers.get("Retry-After")
+            if retry_after:
+                sleep_time = int(retry_after)
+            else:
+                # Check for RateLimit-Reset header (Unix timestamp)
+                reset_time = response.headers.get("RateLimit-Reset")
+                if reset_time:
+                    import time
+
+                    sleep_time = max(0, int(reset_time) - int(time.time()))
+                else:
+                    # Default to 60 seconds if no headers found
+                    sleep_time = 60
+
+            self._logger.warning(
+                f"Rate limit exceeded. Sleeping for {sleep_time} seconds. "
+                f"Headers: {dict(response.headers)}"
+            )
+            await self._sleeps.sleep(sleep_time)
+            return True
+        return False
+
+    async def _execute_graphql(self, query, variables=None):
+        """Execute a GraphQL query with rate limit handling.
+
+        Args:
+            query (str): GraphQL query string
+            variables (dict, optional): Query variables
+
+        Returns:
+            dict: GraphQL response data
+        """
+        session = self._get_session()
+        payload = {"query": query}
+        if variables:
+            payload["variables"] = variables
+
+        async with session.post(self.graphql_url, json=payload) as response:
+            # Handle rate limiting - sleeps and raises exception for @retryable to catch
+            if await self._handle_rate_limit(response):
+                msg = "Rate limit exceeded"
+                raise Exception(msg)
+
+            response.raise_for_status()
+            result = await response.json()
+
+            # Check for GraphQL errors
+            if "errors" in result:
+                errors = result["errors"]
+                # Check if it's a rate limit error in GraphQL response
+                for error in errors:
+                    error_msg = str(error.get("message", "")).lower()
+                    if "rate" in error_msg and "limit" in error_msg:
+                        self._logger.warning(f"GraphQL rate limit error: {error}")
+                        await self._sleeps.sleep(60)
+                        msg = "GraphQL rate limit exceeded"
+                        raise Exception(msg)
+
+                # If not rate limit, raise the error
+                error_msg = f"GraphQL errors: {errors}"
+                raise Exception(error_msg)
+
+            return result.get("data", {})
+
+    async def _get_rest(self, endpoint, params=None):
+        """Execute a REST GET request with rate limit handling.
+
+        Args:
+            endpoint (str): API endpoint path
+            params (dict, optional): Query parameters
+
+        Returns:
+            dict: JSON response
+        """
+        session = self._get_session()
+        url = f"{self.api_url}/{endpoint}"
+
+        async with session.get(url, params=params) as response:
```

Comment:
We should definitely log this if there are no secrets/PII here

Reply:
Done.